pkg/sparrow/controller.go

```diff
@@ -0,0 +1,270 @@
+
+package sparrow
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/caas-team/sparrow/internal/logger"
+	"github.com/caas-team/sparrow/pkg/checks"
+	"github.com/caas-team/sparrow/pkg/checks/runtime"
+	"github.com/caas-team/sparrow/pkg/db"
+	"github.com/caas-team/sparrow/pkg/factory"
+	"github.com/getkin/kin-openapi/openapi3"
+)
+
+// ChecksController is responsible for managing checks.
+type ChecksController struct {
+	db      db.DB
+	metrics Metrics
+	checks  runtime.Checks
+	cErr    chan error
+	done    chan struct{}
+}
+
+// NewChecksController creates a new ChecksController.
+func NewChecksController(dbase db.DB, metrics Metrics) *ChecksController {
+	return &ChecksController{
+		db:      dbase,
+		metrics: metrics,
+		checks:  runtime.Checks{},
+		cErr:    make(chan error, 1),
+		done:    make(chan struct{}, 1),
+	}
+}
+
+// HandleErrors listens for errors and unregisters failed checks.
+func (cc *ChecksController) HandleErrors(ctx context.Context) {
+	log := logger.FromContext(ctx)
+
+	for {
+		select {
+		case err := <-cc.cErr:
+			var runErr *ErrRunningCheck
+			if errors.As(err, &runErr) {
+				uErr := cc.UnregisterCheck(ctx, runErr.Check)
+				if uErr != nil {
+					log.ErrorContext(ctx, "Failed to unregister check", "check", runErr.Check.Name(), "error", uErr)
+				}
+			}
+			log.ErrorContext(ctx, "Error while running check", "error", err)
+		case <-ctx.Done():
+			return
+		case <-cc.done:
+			cc.cErr <- nil
+			return
+		}
+	}
+}
+
+// Shutdown shuts down the ChecksController.
+func (cc *ChecksController) Shutdown(ctx context.Context) (err error) {
+	log := logger.FromContext(ctx)
+
+	for _, c := range cc.checks.Iter() {
+		cErr := cc.UnregisterCheck(ctx, c)
+		if cErr != nil {
+			log.Error("Failed to unregister check while shutting down", "error", cErr)
+			err = errors.Join(err, cErr)
+		}
+	}
+	cc.done <- struct{}{}
+	close(cc.done)
+	return err
+}
+
+// Reconcile reconciles the checks.
+// It registers new checks, updates existing checks and unregisters checks not in the new config.
+func (cc *ChecksController) Reconcile(ctx context.Context, cfg runtime.Config) {
+	log := logger.FromContext(ctx)
+
+	newChecks, err := factory.NewChecksFromConfig(cfg)
+	if err != nil {
+		log.ErrorContext(ctx, "Failed to create checks from config", "error", err)
+		return
+	}
+
+	// Update existing checks and create a list of checks to unregister
+	var unregList []checks.Check
+	for _, c := range cc.checks.Iter() {
+		conf := cfg.For(c.Name())
+		if conf != nil {
+			err = c.SetConfig(conf)
+			if err != nil {
+				log.ErrorContext(ctx, "Failed to set config for check", "check", c.Name(), "error", err)
+			}
+			delete(newChecks, c.Name())
+		} else {
+			unregList = append(unregList, c)
+		}
+	}
+
+	// Unregister checks not in the new config
+	for _, c := range unregList {
+		err = cc.UnregisterCheck(ctx, c)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to unregister check", "check", c.Name(), "error", err)
+		}
+	}
```

Comment:
Can you elaborate why this unreglist is created, instead of directly Unregistering in the Update loop? Is there a technical reason or is this just for simplicity?

Reply:
The `unregList` is to avoid modifying the `cc.checks` slice while iterating over it. This can cause issues because it's generally not safe to modify a slice while iterating over it because it could lead to a runtime panic. Here a simple example of it: https://go.dev/play/p/uzLeQrqb0_A?v=goprev.go